core/src/main/scala/chisel3/Aggregate.scala

```diff
@@ -318,6 +346,190 @@ sealed class Vec[T <: Data] private[chisel3] (gen: => T, val length: Int)
     }
     curLayer(0)
   }
+
+  /** Creates a Vec literal of this type with specified values. this must be a chisel type.
+    *
+    * @param elementInitializers literal values, specified as a pair of the Vec field to the literal value.
+    * The Vec field is specified as a function from an object of this type to the field.
+    * Fields that aren't initialized to DontCare, and assignment to a wire will overwrite any
+    * existing value with DontCare.
+    * @return a Vec literal of this type with subelement values specified
+    *
+    * @example {{{
+    * class MyVec extends Vec {
+    *   val a = UInt(8.W)
+    *   val b = Bool()
+    * }
+    *
+    * (mew MyVec).Lit(
+    *   _.a -> 42.U,
+    *   _.b -> true.B
+    * )
+    * }}}
+    */
+  private[chisel3] def _makeLit(elementInitializers: (Int, Data)*): this.type = {
+
+    // Returns pairs of all fields, element-level and containers, in a Record and their path names
+    def getRecursiveFields(data: Data, path: String): Seq[(Data, String)] = data match {
+      case data: Record =>
+        data.elements.map { case (fieldName, fieldData) =>
+          getRecursiveFields(fieldData, s"$path.$fieldName")
+        }.fold(Seq(data -> path)) {
+          _ ++ _
+        }
+      case data: Vec[_] =>
+        data.getElements.zipWithIndex.map { case (fieldData, fieldIndex) =>
+          getRecursiveFields(fieldData, path = s"$path($fieldIndex)")
+        }.fold(Seq(data -> path)) {
+          _ ++ _
+        }
+      case data => Seq(data -> path) // we don't support or recurse into other Aggregate types here
+    }
+
+    // Returns pairs of corresponding fields between two Vecs of the same type and size
+    def getMatchedFields(x: Data, y: Data): Seq[(Data, Data)] = (x, y) match {
+      case (x: Element, y: Element) =>
+        require(x typeEquivalent y)
+        Seq(x -> y)
+      case (x: Record, y: Record) =>
+        (x.elements zip y.elements).map { case ((xName, xElt), (yName, yElt)) =>
+          require(xName == yName) // assume fields returned in same, deterministic order
+          getMatchedFields(xElt, yElt)
+        }.fold(Seq(x -> y)) {
+          _ ++ _
+        }
+      case (x: Vec[_], y: Vec[_]) =>
+        (x.getElements zip y.getElements).map { case (xElt, yElt) =>
+          getMatchedFields(xElt, yElt)
+        }.fold(Seq(x -> y)) {
+          _ ++ _
+        }
+    }
+
+    def checkLiteralConstruction(): Unit = {
+      val dupKeys = elementInitializers.map { x => x._1 }.groupBy(x => x).flatMap { case (k, v) =>
```

Comment:
There are shorter ways to detect duplicates, comparing size against `.distinct` or `.toSet`
https://stackoverflow.com/questions/7691971/easiest-way-to-decide-if-list-contains-duplicates

More set operations can probably get the duplicated keys.

You don't get the count, but I'd think the index is the important aspect.